File: DataStructureGo/linkedlist.go
package main

import "fmt"

type ListNode struct {
	value interface{}
	next  *ListNode
}

type LinkedList struct {
	head *ListNode
}

func MakeLinkedList() LinkedList {
	return LinkedList{nil}
}

func (list *LinkedList) print() {
	tmp := list.head
	for tmp != nil {
		fmt.Printf("%d, ", tmp.value)
		tmp = tmp.next
	}
	fmt.Print("\n")
}

func (list *LinkedList) addFirst(item interface{}) {
	tmp := list.head
	node := ListNode{item, tmp}
	list.head = &node
}

func (list *LinkedList) addLast(item interface{}) {
	if list.head == nil {
		list.addFirst(item)
		return
	}

	tmp := list.head
	for tmp.next != nil {
		tmp = tmp.next
	}
	tmp.next = &ListNode{item, nil}
}

func (list *LinkedList) insertAfter(key interface{}, item interface{}) {
	tmp := list.head
	for tmp != nil && tmp.value != key {
		tmp = tmp.next
	}
	tmp.next = &ListNode{item, tmp.next}
}

func (list *LinkedList) insertBefore(key interface{}, item interface{}) {
	if list.head == nil {
		return
	}
	if list.head.value == key {
		list.addFirst(item)
		return
	}
	var prev *ListNode
	curr := list.head
	for curr != nil && curr.value != key {
		prev = curr
		curr = curr.next
	}
	if curr != nil {
		prev.next = &ListNode{item, curr}
	}
}

func (list *LinkedList) remove(key interface{}) {
	if list.head == nil {
		return
	}
	if list.head.value == key {
		list.head = list.head.next
		return
	}
	var prev *ListNode
	curr := list.head
	for curr != nil && curr.value != key {
		prev = curr
		curr = curr.next
	}
	if curr != nil {
		prev.next = curr.next
	}
}

File: DataStructureCSharp/DataStructure/LinkedList.cs
public class LinkedList<T> where T: IEquatable<T> {

    public Node<T>? head { get; private set; }

    public LinkedList() {
        head = null;
    }

    public class Node<T> where T: IEquatable<T> {

        public T data { get; internal set; }
        public Node<T>? next { get; internal set; }

        public Node(T data, Node<T>? next) {
            this.data = data;
            this.next = next;
        }
    }

    public void AddFirst(T item) {
        head = new Node<T>(item, null);
    }

    public void AddLast(T item) {
        if (head == null) {
            AddFirst(item);
            return;
        }

        var tmp = head;
        while (tmp.next != null) {
            tmp = tmp.next;
        }
        tmp.next = new Node<T>(item, null);
    }

    public void InsertBefore(T key, T value) {
        if (head == null) {
            return;
        }

        if (head.data.Equals(key)) {
            AddFirst(value);
            return;
        }

        Node<T>? prev = null;
        Node<T>? curr = head;
        while (curr != null && !curr.data.Equals(key)) {
            prev = curr;
            curr = curr.next;
        }

        if (prev != null && curr != null) {
            prev.next = new Node<T>(value, curr);
        }
    }

    public void Remove(T key) {
        if (head == null) {
            return;
        }

        if (head.data.Equals(key)) {
            head = head.next;
            return;
        }

        Node<T>? prev = null;
        Node<T>? curr = head;

        while (curr != null && !curr.data.Equals(key)) {
            prev = curr;
            curr = curr.next;
        }

        if (prev != null && curr != null) {
            prev.next = curr.next;
        }
    }

    public override string ToString() {
        var res = "LinkedList: ";
        if (head == null) {
            res += "null";
            return res;
        }

        var curr = head;
        while (curr != null) {
            res += $"{curr.data}";
            if (curr.next != null) {
                res += " -> ";
            }
            curr = curr.next;
        }
        return res;
    }

}

